azure/azure_kv.go

```diff
@@ -0,0 +1,185 @@
+package azure
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest/to"
+
+	"github.com/Azure/azure-sdk-for-go/services/keyvault/2016-10-01/keyvault"
+	"github.com/libopenstorage/secrets"
+)
+
+const (
+	Name       = "azure-kv"
+	AzureCloud = "AzurePublicCloud"
+	// AzureTenantID for Azure Active Directory
+	AzureTenantID = "AZURE_TENANT_ID"
+	// AzureClientID of service principal account
+	AzureClientID = "AZURE_CLIENT_ID"
+	// AzureClientSecret of service principal account
+	AzureClientSecret = "AZURE_CLIENT_SECRET"
+	// AzureEnviornment to connect
+	AzureEnviornment = "AZURE_ENVIORNMENT"
+	// AzureVaultURI of azure key vault
+	AzureVaultURL = "AZURE_VAULT_URL"
+)
+
+var (
+	ErrAzureTenantIDNotSet    = errors.New("AZURE_TENANT_ID not set.")
+	ErrAzureClientIDNotSet    = errors.New("AZURE_CLIENT_ID not set.")
+	ErrAzureSecretIDNotSet    = errors.New("AZURE_SECRET_ID not set.")
+	ErrAzureVaultURLNotSet    = errors.New("AZURE_VAULT_URL not set.")
+	ErrAzureEnvironmentNotset = errors.New("AZURE_ENVIRONMENT not set.")
+	ErrAzureConfigMissing     = errors.New("AzureConfig is not provided")
+	ErrAzureAuthentication    = errors.New("Azure authentication failed")
+)
+
+type azureSecrets struct {
+	kv      keyvault.BaseClient
+	baseURL string
+}
+
+func New(
+	secretConfig map[string]interface{},
+) (secrets.Secrets, error) {
+
+	tenantID := getAzureKVParams(secretConfig, "AZURE_TENANT_ID")
+	if tenantID == "" {
+		return nil, ErrAzureTenantIDNotSet
+	}
+	clientID := getAzureKVParams(secretConfig, "AZURE_CLIENT_ID")
+	if clientID == "" {
+		return nil, ErrAzureClientIDNotSet
+	}
+	secretID := getAzureKVParams(secretConfig, "AZURE_CLIENT_SECRET")
+	if secretID == "" {
+		return nil, ErrAzureSecretIDNotSet
+	}
+	envName := getAzureKVParams(secretConfig, "AZURE_ENVIORNMENT")
+	if envName == "" {
+		// we set back to default AzurePublicCloud
+		envName = AzureCloud
+	}
+	vaultURL := getAzureKVParams(secretConfig, "AZURE_VAULT_URL")
+	if vaultURL == "" {
+		return nil, ErrAzureVaultURLNotSet
+	}
+
+	client, err := getAzureVaultClient(clientID, secretID, tenantID, envName)
+	if err != nil {
+		return nil, ErrAzureAuthentication
+	}
+
+	return &azureSecrets{
+		kv:      client,
+		baseURL: vaultURL,
+	}, nil
+}
+
+func (az *azureSecrets) GetSecret(
+	secretID string,
+	keyContext map[string]string,
+) (map[string]interface{}, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 6000*time.Second)
+	defer cancel()
+
+	if secretID == "" {
+		return nil, secrets.ErrEmptySecretId
+	}
+	secretResp, err := az.kv.GetSecret(ctx, az.baseURL, secretID, "")
+	if err != nil {
+		return nil, err
+	}
+
+	secretData := make(map[string]interface{})
+	err = json.Unmarshal([]byte(*secretResp.Value), &secretData)
```

Comment:
On a safe side check `secretResp.Value` is not nil, else the code will panic

Reply:
Should be captured by GetSecret() call, but to be on safe side added check for same

Reply:
Right, but was thinking if someone sets a empty value from azure side and gives us the key to use. Anyway, it's better to have nil pointer checks.